File: packages/ts/tests/zyins/licenses.test.ts
import { describe, expect, it } from 'vitest';
import { ZyInsError } from '../../src/zyins';
import { client, recordingTransport } from './client-test-helpers';

// jsonObject parses a JSON object body for assertion. Defensive: returns
// an empty object when the body is not an object, so tests fail with
// readable expectation diffs instead of typecast errors.
function jsonObject(raw: string): Record<string, unknown> {
  const parsed: unknown = JSON.parse(raw);
  if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(parsed)) {
      out[k] = v;
    }
    return out;
  }
  return {};
}

describe('ZyInsClient.licenses.activate', () => {
  const ACTIVATE_OK = JSON.stringify({
    status: 'active',
    remainingActivations: 2,
    auth: { licenseKey: 'LIC-NEW-7890' },
  });

  it('POSTs JSON to /v1/licenses/activate with email + keycode + device_id', async () => {
    const { transport, requests } = recordingTransport(200, ACTIVATE_OK);
    const c = client(transport);
    const result = await c.licenses.activate({
      email: 'john.doe@acme-agency.com',
      keycode: 'ABC-123-XYZ',
      deviceId: 'device-xyz-123',
    });
    expect(result.status).toBe('active');
    expect(result.remainingActivations).toBe(2);
    expect(result.auth.licenseKey).toBe('LIC-NEW-7890');
    expect(requests).toHaveLength(1);
    expect(requests[0]!.method).toBe('POST');
    expect(requests[0]!.url).toBe('https://test.example/v1/licenses/activate');
    expect(requests[0]!.headers['Content-Type']).toBe('application/json');
    expect(requests[0]!.headers['Idempotency-Key']).toBeTruthy();
    const payload = jsonObject(requests[0]!.body);
    expect(payload['email']).toBe('john.doe@acme-agency.com');
    expect(payload['keycode']).toBe('ABC-123-XYZ');
    expect(payload['device_id']).toBe('device-xyz-123');
  });

  it('accepts the ADR-012 envelope shape', async () => {
    const { transport } = recordingTransport(
      200,
      JSON.stringify({
        data: { status: 'active', remainingActivations: 0, auth: { licenseKey: 'LK-1' } },
      }),
    );
    const c = client(transport);
    const result = await c.licenses.activate({
      email: 'x@x',
      keycode: 'ABC-123-XYZ',
      deviceId: 'd',
    });
    expect(result.auth.licenseKey).toBe('LK-1');
    expect(result.remainingActivations).toBe(0);
  });

  it('rejects missing deviceId', async () => {
    const { transport } = recordingTransport(200, ACTIVATE_OK);
    const c = client(transport);
    await expect(
      c.licenses.activate({ email: 'x@x', keycode: 'ABC-123-XYZ', deviceId: '' }),
    ).rejects.toThrow(/deviceId/);
  });

  it('rejects malformed success responses', async () => {
    const { transport } = recordingTransport(200, JSON.stringify({ status: 'active' }));
    const c = client(transport);
    await expect(
      c.licenses.activate({ email: 'x@x', keycode: 'ABC-123-XYZ', deviceId: 'd' }),
    ).rejects.toThrow(/remainingActivations/);
  });

  it('surfaces 4xx as a typed error', async () => {
    const { transport } = recordingTransport(
      400,
      JSON.stringify({ type: 'about:blank', title: 'bad request', status: 400, code: 'validation_error' }),
    );
    const c = client(transport);
    const promise = c.licenses.activate({ email: 'x@x', keycode: 'ABC-123-XYZ', deviceId: 'd' });
    await expect(promise).rejects.toBeInstanceOf(ZyInsError);
    await expect(promise).rejects.toMatchObject({ code: 'validation_error', httpStatus: 400 });
  });

  it('surfaces 500 as a typed error', async () => {
    const { transport } = recordingTransport(
      500,
      JSON.stringify({ type: 'about:blank', title: 'server error', status: 500, code: 'server_error' }),
    );
    const c = client(transport);
    const promise = c.licenses.activate({ email: 'x@x', keycode: 'ABC-123-XYZ', deviceId: 'd' });
    await expect(promise).rejects.toBeInstanceOf(ZyInsError);
    await expect(promise).rejects.toMatchObject({ code: 'unknown', httpStatus: 500 });
  });
});

describe('ZyInsClient.licenses.check', () => {
  it('POSTs JSON to /v1/licenses/check with email + keycode', async () => {
    const { transport, requests } = recordingTransport(200, JSON.stringify({ status: 'valid' }));
    const c = client(transport);
    const result = await c.licenses.check({
      email: 'john.doe@acme-agency.com',
      keycode: 'ABC-123-XYZ',
      deviceId: 'device-1',
    });
    expect(result.status).toBe('valid');
    expect(requests).toHaveLength(1);
    expect(requests[0]!.method).toBe('POST');
    expect(requests[0]!.url).toBe('https://test.example/v1/licenses/check');
    expect(requests[0]!.headers['Content-Type']).toBe('application/json');
    expect(requests[0]!.headers['Idempotency-Key']).toBeTruthy();
    const payload = jsonObject(requests[0]!.body);
    expect(payload['email']).toBe('john.doe@acme-agency.com');
    expect(payload['keycode']).toBe('ABC-123-XYZ');
    expect(payload['device_id']).toBe('device-1');
  });

  it('accepts the ADR-012 envelope shape', async () => {
    const { transport } = recordingTransport(200, JSON.stringify({ data: { status: 'inactive' } }));
    const c = client(transport);
    const result = await c.licenses.check({ email: 'x@x', keycode: 'ABC-123-XYZ' });
    expect(result.status).toBe('inactive');
  });

  it('rejects missing email', async () => {
    const { transport } = recordingTransport(200, JSON.stringify({ status: 'valid' }));
    const c = client(transport);
    await expect(c.licenses.check({ email: '', keycode: 'ABC-123-XYZ' })).rejects.toThrow(/email/);
  });

  it('surfaces 500 as a typed error', async () => {
    const { transport } = recordingTransport(
      500,
      JSON.stringify({ type: 'about:blank', title: 'server error', status: 500, code: 'server_error' }),
    );
    const c = client(transport);
    await expect(c.licenses.check({ email: 'x@x', keycode: 'ABC-123-XYZ' })).rejects.toBeInstanceOf(Error);
  });
});

describe('ZyInsClient.licenses.deactivate', () => {
  it('POSTs JSON to /v1/licenses/deactivate and returns the status', async () => {
    const { transport, requests } = recordingTransport(
      200,
      JSON.stringify({ status: 'deactivated' }),
    );
    const c = client(transport);
    const result = await c.licenses.deactivate({
      email: 'john.doe@acme-agency.com',
      keycode: 'ABC-123-XYZ',
    });
    expect(result.status).toBe('deactivated');
    expect(requests[0]!.url).toBe('https://test.example/v1/licenses/deactivate');
    expect(requests[0]!.headers['Idempotency-Key']).toBeTruthy();
  });

  it('rejects missing keycode', async () => {
    const { transport } = recordingTransport(200, '{}');
    const c = client(transport);
    await expect(c.licenses.deactivate({ email: 'x@x', keycode: '' })).rejects.toThrow(/keycode/);
  });

  it('rejects malformed success responses', async () => {
    const { transport } = recordingTransport(200, '{}');
    const c = client(transport);
    await expect(c.licenses.deactivate({ email: 'x@x', keycode: 'ABC-123-XYZ' })).rejects.toThrow(/deactivated/);
  });
});

File: zyins/licenses_test.go
package zyins

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

// captureLicensesRequest records the headers and body of one request
// for assertion in tests that exercise the licenses sub-service.
type captureLicensesRequest struct {
	method string
	path   string
	auth   string
	idem   string
	body   []byte
}

func TestLicenses_Check_HappyPath(t *testing.T) {
	var captured captureLicensesRequest
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		captured.method = r.Method
		captured.path = r.URL.Path
		captured.auth = r.Header.Get("Authorization")
		captured.idem = r.Header.Get("Idempotency-Key")
		captured.body, _ = io.ReadAll(r.Body)
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"status":"valid"}`))
	}))
	defer srv.Close()
	c := newTestClient(t, srv)
	result, err := c.License.Check(context.Background(), &LicenseCheckInput{
		Email:      "john.doe@acme-agency.com",
		Keycode:    "ABC-123-XYZ",
		DeviceID:   "device-1",
		LicenseKey: "abc123",
	})
	if err != nil {
		t.Fatalf("Check: %v", err)
	}
	if result.Status != LicenseStatusValid {
		t.Errorf("status = %q, want valid", result.Status)
	}
	if captured.method != http.MethodPost || captured.path != licensesCheckPath {
		t.Errorf("captured %s %s, want POST %s", captured.method, captured.path, licensesCheckPath)
	}
	if !strings.HasPrefix(captured.auth, "Bearer ") {
		t.Errorf("expected Bearer auth, got %q", captured.auth)
	}
	if captured.idem == "" {
		t.Errorf("expected Idempotency-Key header on POST")
	}
	var bodyParsed licensesCheckWireBody
	if err := json.Unmarshal(captured.body, &bodyParsed); err != nil {
		t.Fatalf("body unmarshal: %v", err)
	}
	if bodyParsed.Email != "john.doe@acme-agency.com" || bodyParsed.Keycode != "ABC-123-XYZ" {
		t.Errorf("body fields not propagated: %+v", bodyParsed)
	}
}

func TestLicenses_Check_AcceptsEnvelopedResponse(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"data":{"status":"inactive"}}`))
	}))
	defer srv.Close()
	c := newTestClient(t, srv)
	result, err := c.License.Check(context.Background(), &LicenseCheckInput{
		Email: "x@x", Keycode: "ABC-123-XYZ",
	})
	if err != nil {
		t.Fatalf("Check: %v", err)
	}
	if result.Status != LicenseStatusInactive {
		t.Errorf("status = %q, want inactive", result.Status)
	}
}

func TestLicenses_Check_RejectsMissingEmail(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		t.Fatalf("request must not hit the wire on validation failure")
	}))
	defer srv.Close()
	c := newTestClient(t, srv)
	_, err := c.License.Check(context.Background(), &LicenseCheckInput{Keycode: "ABC-123-XYZ"})
	if err == nil {
		t.Fatalf("expected validation error, got nil")
	}
	var ve *ValidationError
	if !errors.As(err, &ve) {
		t.Fatalf("expected *ValidationError, got %T: %v", err, err)
	}
}

func TestLicenses_Check_NilInput(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
	defer srv.Close()
	c := newTestClient(t, srv)
	_, err := c.License.Check(context.Background(), nil)
	if err == nil {
		t.Fatalf("expected validation error on nil input")
	}
}

func TestLicenses_Check_ServerError(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/problem+json")
		w.WriteHeader(http.StatusInternalServerError)
		_, _ = w.Write([]byte(`{"type":"about:blank","title":"server error","status":500,"code":"server_error"}`))
	}))
	defer srv.Close()
	c := newTestClient(t, srv)
	_, err := c.License.Check(context.Background(), &LicenseCheckInput{
		Email: "x@x", Keycode: "ABC-123-XYZ",
	})
	if err == nil {
		t.Fatalf("expected error from 500 response")
	}
}

func TestLicenses_Deactivate_HappyPath(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != licensesDeactivatePath {
			t.Errorf("path = %q, want %q", r.URL.Path, licensesDeactivatePath)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"status":"deactivated"}`))
	}))
	defer srv.Close()
	c := newTestClient(t, srv)
	result, err := c.License.Deactivate(context.Background(), &LicenseDeactivateInput{
		Email:    "john.doe@acme-agency.com",
		Keycode:  "ABC-123-XYZ",
		DeviceID: "device-1",
	}, WithIdempotencyKey("550e8400-e29b-41d4-a716-446655440000"))
	if err != nil {
		t.Fatalf("Deactivate: %v", err)
	}
	if result.Status != "deactivated" {
		t.Errorf("status = %q, want deactivated", result.Status)
	}
}

func TestLicenses_Deactivate_RejectsMissingKeycode(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
	defer srv.Close()
	c := newTestClient(t, srv)
	_, err := c.License.Deactivate(context.Background(), &LicenseDeactivateInput{Email: "x@x"})
	if err == nil {
		t.Fatalf("expected validation error")
	}
}
